### DISCUSSION TITLE
Erlang Does Not Build on ARM Based Amazon Linux 2023

### DISCUSSION THREAD

**Discussion by CSimpiFoN** [rabbitmq/erlang-rpm#115] [General]
### Describe the bug

When I try to build on ARM based Amazon Linux 2, the following error is dropped:
<code>checking for systemd/sd-daemon.h... no
configure: error: in /tmp/erlang-rpm-25.3/BUILD/otp-OTP-25.3/erts': configure: error: --enable-systemd was given, but test for systemd failed See config.log' for more details
ERROR: /tmp/erlang-rpm-25.3/BUILD/otp-OTP-25.3/erts/configure failed!
./configure: line 370: kill: (-39330) - No such process
D:      waitpid(39138) rc 39138 status 100
D: /bin/sh failed: 100
error: Bad exit status from /tmp/erlang-rpm-25.3/tmp/rpm-tmp.RtzobR (%build)

RPM build errors:
line 54: It's not recommended to have unversioned Obsoletes: Obsoletes: erlang-docbuilder
Bad exit status from /tmp/erlang-rpm-25.3/tmp/rpm-tmp.RtzobR (%build)
ufdio:       1 reads,    17654 total bytes in 0.000011 secs
ufdio:       1 reads,     5707 total bytes in 0.000003 secs
ufdio:       1 reads,    17654 total bytes in 0.000002 secs
make: *** [Makefile:51: erlang] Error 1</code>

### Reproduction steps

1. Download https://github.com/rabbitmq/erlang-rpm/archive/refs/tags/v25.3.tar.gz
2. Untar the archive
3. Change the directory to the unarchived folder
4. make (as per the README)

### Expected behavior

To get Erlang built and ready to be used by RabbitMQ

### Additional context

Want to install and use RabbitMQ on ARM based Amazon Linux 2023 server
<code>cat /etc/os-release 
NAME="Amazon Linux"
VERSION="2023"
ID="amzn"
ID_LIKE="fedora"
VERSION_ID="2023"
PLATFORM_ID="platform:al2023"
PRETTY_NAME="Amazon Linux 2023"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2023"
HOME_URL="https://aws.amazon.com/linux/"
BUG_REPORT_URL="https://github.com/amazonlinux/amazon-linux-2023"
SUPPORT_END="2028-03-01"</code>

**Comment by michaelklishin** [comment]
```
configure: error: --enable-systemd was given, but test for systemd failed See config.log' for more details
```

We do not maintain Erlang/OTP, so either it is a problem with systemd detection in `./configure`, or your system is missing `systemd`, or a portion of it that is used for detection, similarly to [this](https://erlangforums.com/t/erlang-25-0-4-failed-to-compile-in-fedora-release-36/1822) (this is just an example).

We do not maintain Erlang/OTP and don't have the time to investigate Erlang build failures.

**Comment by michaelklishin** [comment]
Some 6-12 months ago someone from (IIRC) Red Hat set up a Fedora Copr builds for ARM64 based on this packages but [official Fedora Erlang images](https://packages.fedoraproject.org/pkgs/erlang/erlang/fedora-37-updates.html#provides) do not list arm64 as a platform.

It could be a [package directory website limitation](https://koji.fedoraproject.org/koji/buildinfo?buildID=2166886). However, for Fedora-based distributions such as Amazon Linux 2023, using
a Fedora community package on ARM64 should be a perfectly fine option.

The Erlang and Elixir packages are up-to-date, and the broader Fedora community is many orders of
magnitude larger than the RabbitMQ core team.

**Reply by michaelklishin** [reply]
I was wrong, https://koji.fedoraproject.org/koji/buildinfo?buildID=2166886 does mention `aarch64` files. So the `erlang` package for Fedora 37 and upcoming 38 is built for ARM64. And they [patch Erlang's `./configure` script](https://koji.fedoraproject.org/koji/rpminfo?rpmID=33854207), although it does not seem to be systemd detection-related, judging from the file names.

**Comment by michaelklishin** [comment]
An attempt to use Fedora packages on Amazon Linux 2023 (ARM64) in Docker, such as

```
dnf install -y https://kojipkgs.fedoraproject.org//packages/erlang/25.3/1.fc37/aarch64/erlang-erts-25.3-1.fc37.aarch64.rpm
```

Lead to a few unfortunate discoveries:

 * AL2023 only provides OpenSSL 3.0, which Erlang 25.3 is technically compatible with but there are nuances
 * AL2023 does not provide Erlang packages in its main repo, so it is not Fedora-based but CentOS-based in practice
 * AL2023 does not satisfy a few Erlang package dependencies

So it is Fedora-based mostly in spirit.

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/erlang-rpm/issues/116

**Comment by michaelklishin** [comment]
With https://github.com/rabbitmq/erlang-rpm/pull/117, this RPM builds successfully on at least one target OS on ARM64: Fedora 38.

**Comment by michaelklishin** [comment]
Here is an example ARM64 RPM produced in Docker on an ARM 64 host. You will need to use #119 or wait for it to be merged (assuming the changes work for Fedora, CentOS Stream as well).

The file is a .zip because GitHub does not allow .rpm file uploads.

[erlang-25.3-1.amzn2023.aarch64.rpm.zip](https://github.com/rabbitmq/erlang-rpm/files/11181910/erlang-25.3-1.amzn2023.aarch64.rpm.zip)

**Comment by michaelklishin** [comment]
[25.3](https://github.com/rabbitmq/erlang-rpm/releases/tag/v25.3) is the first release with aarch64 packages provided in binary form.

As of #119, anyone with an ARM64 host and Docker installed can build aarm64 versions of this package
for CentOS Stream 9, Stream 8, Fedora 38+ and Amazon Linux 2023.

aarch64 packages are not built by automatic pipelines, so I'd have to trigger their builds on my ARM64 host every so often (perhaps not for every patch release).

Most importantly, #119 proves that the package did not have any aarch64 incompatibilities, it was a matter of modernizing the spec a bit and automating Docker builds for non-CentOS Stream distributions.
